### PR TITLE
feat(config)!: add new option `constraintsFiltering`

### LINKED ISSUE #19979: Make release compatibility filtering opt-in
### What would you like Renovate to be able to do?

Default to no release filtering (today only python exact versions is supported) so that we can add more advanced filtering in future.

### If you have any ideas on how this should be implemented, please tell us here.

I'd like to make it an enum in case we want to have different degrees of filtering in future.

Feature name `releaseConstraintFiltering`, default value `"none"`.

Alternative value `"strict"`, and put this code behind an `if (config.releaseConstraintFiltering === 'strict')`: https://github.com/renovatebot/renovate/blob/8e8cbba0e62ac5ccac0a24de295a468c5c2101b8/lib/modules/datasource/index.ts#L398-L419

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes

Adds new option `constraintsFiltering`.

BREAKING CHANGE: PyPI releases will no longer be filtered by default based on `constraints.python` compatibility. To retain existing functionality, set `constraintsFiltering=strict`.

## Context

- Closes: #19979 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

BREAKING CHANGE: Python constraint version filtering will be disabled by default, needs to be opted into using `constraintsFiltering=strict`.

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
[Default Mode](https://github.com/RahulGautamSingh-testing/merge3/commit/f7ad9515bb93cf766619484162da7723f0380fc5): 
[Strict Mode](https://github.com/RahulGautamSingh-testing/merge3/pull/17/commits/e48187da04f51f909d37864993b07c917c64e3c3): <strict>

File: lib/modules/datasource/index.spec.ts [hunk 1]
```diff
@@ -632,6 +632,99 @@ describe('modules/datasource/index', () => {
...
+            depName,
+            defaultRegistryUrls: ['https://foo.bar'],
+          });
+          expect(res).toMatchObject({
+            releases: [{ version: '0.0.1' }, { version: '0.0.2' }],
+          });
+        });
+
+        it('keeps all releases if constraints is set but no value defined for releaseConstraintFiltering', async () => {
+          const registries: RegistriesMock = {
```
Comment (viceice) [inline lib/modules/datasource/index.spec.ts:665]: use satisfies operator now

**Review by rarkins — CHANGES_REQUESTED** [review]
Sorry, I just realized that `constraintsFiltering` would be shorter and better - can you rename?

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -2748,7 +2748,7 @@ In case there is a need to configure them manually, it can be done using this `r
 
 The field supports multiple URLs but it is datasource-dependent on whether only the first is used or multiple.
 
-## releaseConstraintFiltering
+## constraintsFiltering
```
Comment (rarkins) [inline docs/usage/configuration-options.md:2751]: Will require reordering

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -525,6 +525,16 @@ If you need to _override_ constraints that Renovate detects from the repository,
 !!! note
     Make sure not to mix this up with the term `compatibility`, which Renovate uses in the context of version releases, e.g. if a Docker image is `node:12.16.0-alpine` then the `-alpine` suffix represents `compatibility`.
 
+## constraintsFiltering
+
+This option allows you to select the type of release filtering you want. Currently we have these two options:
```
Comment (rarkins) [inline docs/usage/configuration-options.md:530]:
```suggestion
This option controls whether Renovate filters new releases based on configured or detected `constraints`.
Renovate supports two options:
```

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -525,6 +525,16 @@ If you need to _override_ constraints that Renovate detects from the repository,
...
+`none`: No release filtering
+`strict`: If any of the release's constraints match, then it's included
```
Comment (rarkins) [inline docs/usage/configuration-options.md:532]:
```suggestion
 - `none`: No release filtering (all releases allowed)
 - `strict`: If the release's constraints match the package file constraints, then it's included
```

File: docs/usage/configuration-options.md [hunk 5]
```diff
@@ -525,6 +525,16 @@ If you need to _override_ constraints that Renovate detects from the repository,
...
+
+We are working on adding more advanced filtering options.
+
+Note: There must be a `constraints` object present in your renovate config for this to work.
+
```
Comment (rarkins) [inline docs/usage/configuration-options.md:537]: This feature is limited to `pypi` datasource only.

File: lib/config/options/index.ts [hunk 6]
```diff
@@ -217,6 +217,16 @@ const options: RenovateOptions[] = [
     cli: false,
     env: false,
   },
+  {
+    name: 'constraintsFiltering',
+    description:
+      'This option allows you to select the type of release filtering.',
```
Comment (rarkins) [inline lib/config/options/index.ts:223]:
```suggestion
      'Perform release filtering based on language constraints.',
```

File: lib/config/options/index.ts [hunk 7]
```diff
@@ -217,6 +217,16 @@ const options: RenovateOptions[] = [
...
+    type: 'string',
+    allowedValues: ['none', 'strict'],
+    stage: 'repository',
+    default: 'none',
+    experimental: true,
```
Comment (rarkins) [inline lib/config/options/index.ts:228]:
```suggestion
```

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -217,6 +217,16 @@ const options: RenovateOptions[] = [
     cli: false,
     env: false,
   },
+  {
+    name: 'constraintsFiltering',
+    description:
+      'This option allows you to select the type of release filtering.',
+    type: 'string',
+    allowedValues: ['none', 'strict'],
+    stage: 'repository',
```
Comment (rarkins) [inline lib/config/options/index.ts:226]: I'm not sure about this. It's used during lookups

File: lib/config/types.ts [hunk 9]
```diff
@@ -188,6 +188,7 @@ export interface RegExManager extends RegexManagerTemplates {
 }
 
 export type UseBaseBranchConfigType = 'merge' | 'none';
+export type ReleaseConstraintFilter = 'strict' | 'none';
```
Comment (rarkins) [inline lib/config/types.ts:191]:
```suggestion
export type ConstraintsFilter = 'strict' | 'none';
```

Comment (rarkins) [reply lib/config/types.ts:191]: You will need to update all references

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 5 above)
Comment (RahulGautamSingh) [reply docs/usage/configuration-options.md]: Should I use admonitions here?

!!! note 
- <listItem 1>
- <listItem 2>

image

Comment (RahulGautamSingh) [reply docs/usage/configuration-options.md]: I am not sure if this list supported. I just made this using inspect element to show an example :)
Also there are many cases where we don't us admonitions for `Note:`, is htere a reason or they are pending?

**Review by viceice — APPROVED** [review]

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -525,6 +525,19 @@ If you need to _override_ constraints that Renovate detects from the repository,
...
+This option controls whether Renovate filters new releases based on configured or detected `constraints`.
+Renovate supports two options:
+
+- `none`: No release filtering (all releases allowed)
+- `strict`: If the release's constraints match the package file constraints, then it's included
+
+We are working on adding more advanced filtering options.
+
+Note: There must be a `constraints` object present in your renovate config for this to work.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:538]:
```suggestion
Note: There must be a `constraints` object in your Renovate config for this to work.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Pull request merged**